docs/adrs/0003-bulk-import-ui.md

```diff
@@ -0,0 +1,98 @@
+* Selecting several data files to concurrently import into new data objects.
+* Launching what would amount to several import jobs at once (i.e. a bulk import operation).
+* Monitoring the state of those jobs and restarting if necessary.
+* Viewing the results of those jobs.
+
+Currently, this work is done by users using one cell at a time. An App Cell is generated for each desired data file upload, which can be time-consuming and cluttering for Narratives. The new design to support bulk operations.
+
+## Author
+
+@briehl
+
+## Status
+
+In progress
+
+## Background
+
+App Cells (and other cells) are all built using the [Jupyter front end extension framework](https://jupyter-notebook.readthedocs.io/en/stable/extending/frontend_extensions.html). On application and page load, this framework allows the overriding and customization of most FE components. Specifically, in the Narrative, this framework is only used for Cells. 
+
+Jupyter Notebook (and thus Narrative) cells come in three types - Code, Markdown, and Raw. All KBase cell extensions operate on Jupyter Code cells. It's easiest to think of them as subclassing - an App Cell is a Code Cell specialized for running Narrative Apps. Each of these extensions is activated by creating a cell with a specific metadata component that describes how the cell should be treated. For example, adding the following block to code cell metadata tells the extension machinery to treat it as a KBase App Cell:
+```
+{
+    kbase: {
+        type: 'app'
+    }
+}
+```
+
+The App Cell is, by far, the most complicated extension cell type. It has several components that all serve different purposes:
+* A finite state machine (FSM) that's used to determine cell state.
+* A core system used to render the cell based on state.
+* A message bus that's used to communicate with the rest of the FE application, as well as the BE (kernel) to get job state.
+* An interpreter / manager for app specs - turns the app specs into sets of input form elements.
+* A code translator - turns the various sets of inputs into code that can be executed to run the App.
+
+The reason for the App cell being an extension to code cells is the last point above: in essence, the primary UI components of the App cell serve as a fancy code generator. User interactions with form elements generate the code that can be used to start apps and view their results. Thus, anything that an app cell does with the UI can be done programmatically by an advanced user.
+
+> It should also be noted here that, regardless of the choice made, we may need to alter how app specs are made, interpreted, and rendered. The bulk import design has various inputs and outputs listed in horizontal rows while the existing app specs are rendered vertically. There's no notion of rows/columns in the app spec right now or how to visually display various inputs along a grid pattern. In order to effectively match the design we may need to do one of the following:
```

Comment:
A bit tangential to the ADR specifically but if there's any suggestions we have to make the design less onerous on whatever development path we choose we should consider and discuss with design.

Reply:
Oh nm!  I finally saw the resources link at the bottom 🤦